File: hasel2/go/main.go
package main

import (
    "fmt"
    "os"
    "strings"
)

func main() {
    input := os.Getenv("HASEL_INPUT")
    if input == "" {
        input = "Hasel λ🐿 2.0"
    }
    if input == "" && len(os.Args) > 1 {
        input = strings.Join(os.Args[1:], " ")
    }
    if input == "" {
        input = "Hasel λ🐿 2.0"
    }

    emit("input", input)
    emit("everything", "*"+input+"*")
    emit("regex", ".*"+regexEscape(input)+".*")
    emit("ascii_printable", asciiPrintable(input))
    emit("contains_non_ascii", boolString(containsNonASCII(input)))
    emit("tokens", strings.Join(strings.Fields(strings.ToLower(input)), ","))
    emit("codepoints", codepoints(input))
}

func emit(key string, value string) {
    fmt.Printf("%s\t%s\n", key, value)
}

func boolString(v bool) string {
    if v {
        return "true"
    }
    return "false"
}

func containsNonASCII(input string) bool {
    for _, r := range input {
        if r > 127 {
            return true
        }
    }
    return false
}

func asciiPrintable(input string) string {
    out := strings.Builder{}
    for _, r := range input {
        if r == ' ' || (r >= 33 && r <= 126) {
            out.WriteRune(r)
        } else {
            out.WriteRune('?')
        }
    }
    return out.String()
}

func regexEscape(input string) string {
    meta := `\\.^$|?*+()[]{}'
    out := strings.Builder{}
    for _, r := range input {
        if strings.ContainsRune(meta, r) {
            out.WriteRune('\\')
        }
        out.WriteRune(r)
    }
    return out.String()
}

func codepoints(input string) string {
    parts := make([]string, 0, len(input))
    for _, r := range input {
        parts = append(parts, fmt.Sprintf("U+%04X", r))
    }
    return strings.Join(parts, ",")
}

File: hasel2/haskell/Main.hs
module Main where

import Data.Char (isAscii, isPrint, ord, toLower, toUpper)
import Data.List (intercalate)
import Numeric (showHex)
import System.Environment (getArgs, lookupEnv)
import System.IO (hSetEncoding, stderr, stdout, utf8)

main :: IO ()
main = do
  hSetEncoding stdout utf8
  hSetEncoding stderr utf8
  args <- getArgs
  envInput <- lookupEnv "HASEL_INPUT"
  let input = resolveInput envInput args
  emitReport input

defaultInput :: String
defaultInput = "Hasel λ🐿 2.0"

resolveInput :: Maybe String -> [String] -> String
resolveInput (Just value) _ = value
resolveInput Nothing [] = defaultInput
resolveInput Nothing xs = unwords xs

emitReport :: String -> IO ()
emitReport input = do
  emit "input" input
  emit "everything" ("*" ++ input ++ "*")
  emit "regex" (".*" ++ regexEscape input ++ ".*")
  emit "ascii_printable" (map asciiPrintable input)
  emit "contains_non_ascii" (boolString (any (not . isAscii) input))
  emit "tokens" (intercalate "," (words (map toLower input)))
  emit "codepoints" (intercalate "," (map formatCodePoint input))

emit :: String -> String -> IO ()
emit key value = putStrLn (key ++ "\t" ++ value)

boolString :: Bool -> String
boolString True = "true"
boolString False = "false"

asciiPrintable :: Char -> Char
asciiPrintable ch
  | ch == ' ' = ' '
  | isAscii ch && isPrint ch = ch
  | otherwise = '?'

regexEscape :: String -> String
regexEscape = concatMap escape
  where
    escape ch
      | ch `elem` regexMeta = ['\\', ch]
      | otherwise = [ch]
    regexMeta = "\\.^$|?*+()[]{}"

formatCodePoint :: Char -> String
formatCodePoint ch = "U+" ++ pad4 (map toUpper (showHex (ord ch) ""))

pad4 :: String -> String
pad4 raw = replicate (max 0 (4 - length raw)) '0' ++ raw
